File: Java/138. 复制带随机指针的链表.java
/*
执行用时：0 ms, 在所有 Java 提交中击败了100.00% 的用户
内存消耗：37.9 MB, 在所有 Java 提交中击败了74.47% 的用户
*/
class Solution {
    public Node copyRandomList(Node head) {
        copyList(head);
        copyRandomPtr(head);
        return seperateList(head);    
    }
    private void copyList(Node head){
        Node cursor = head;
        Node next = null;
        while(cursor != null){
            next = cursor.next;
            cursor.next = new Node(cursor.val);
            cursor.next.next = next;
            cursor = next;
        }
    }
    private void copyRandomPtr(Node head){
        Node cursor = head;
        Node copy = null;
        while(cursor!=null){
            copy = cursor.next;
            if(cursor.random != null){
                cursor.next.random = cursor.random.next;
            }
            cursor = copy.next;
        }

    }
    private Node seperateList(Node head){
        Node result = new Node(0);
        Node cursor = result;
        Node copy = null;
        while(head != null){
            copy = head.next;
            head.next = copy.next;
            cursor.next = copy;
            head = head.next;
            cursor = cursor.next;
        }
        return result.next;
    }
}

File: CPlusPlus/138. 复制带随机指针的链表.cpp
#include "include/linked_list.hpp"
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head == nullptr) return nullptr;
        head = copy_list(head);
        copy_rand(head);
        return split(head);
    }
private:
    Node* copy_list(Node* head){
        Node* cursor = head;
        while(cursor){
            Node* temp = new Node(cursor->val);
            temp ->next = cursor->next;
            cursor->next = temp;
            cursor = cursor->next->next;
        }
        return head;
    }
    void copy_rand(Node* head){
        while(head && head->next){
            if(head->random){
                head->next->random = head->random->next;
            }
            head = head->next->next;
        }
    }
    Node* split(Node* head){
        Node* res = head->next;
        Node* curhead = head;
        Node* curres = res;
        while(curres && curhead){
            curhead->next = curres->next;
            curhead = curhead->next;
            if(curhead) curres->next = curhead->next;
            curres = curres->next;
        }
        return res;
    }
};
